core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -881,6 +881,7 @@ private void initialize() {
 
         TopicData<PartitionAllData> state = response.topicsData().get(0);
         if (state.topicId() != topicIdPartition.topicId() || state.partitions().size() != 1
+            || state.partitions().get(0).errorCode() != Errors.NONE.code()
```

Comment:
The following are the error codes from ReadShareGroupStateResponse. It seems that we need to treat them differently.

  // - NOT_COORDINATOR (version 0+) 
  // - COORDINATOR_NOT_AVAILABLE (version 0+)
  // - COORDINATOR_LOAD_IN_PROGRESS (version 0+)
  // - GROUP_ID_NOT_FOUND (version 0+)
  // - UNKNOWN_TOPIC_OR_PARTITION (version 0+)
  // - FENCED_LEADER_EPOCH (version 0+)
  // - INVALID_REQUEST (version 0+)

NOT_COORDINATOR, COORDINATOR_NOT_AVAILABLE, COORDINATOR_LOAD_IN_PROGRESS: These are transient errors. So we probably want to send a retriable error in the fetch response. I don't see a suitable error code listed on the KIP. Perhaps we need to add one?

GROUP_ID_NOT_FOUND: We probably should just return INVALID_GROUP_ID and add it in the error codes for ShareFetchResponse in the KIP?

UNKNOWN_TOPIC_OR_PARTITION: We should just return the UNKNOWN_TOPIC_OR_PARTITION error.

INVALID_REQUEST: We should just return INVALID_REQUEST.

FENCED_LEADER_EPOCH: This is also transient. So we need to return a retriable error. We also need to remove this SharePartition so that a new one with a new leader epoch could be created. Otherwise, we will be getting this error forever.

What do you think, @AndrewJSchofield ?

Reply:
@junrao sounds about right. I'll talk to @apoorvmittal10 about the details.

Reply:
@junrao I discussed with @apoorvmittal10 and @smjn.

NOT_COORDINATOR, COORDINATOR_NOT_AVAILABLE and COORDINATOR_LOAD_IN_PROGRESS are transient errors that are intended to be handled automatically by the persister which makes the calls to the ReadShareGroupState RPC. The persister will retry the RPCs with exponential back-off and we do not expect these error codes to surface to the SharePartition. The retries do have a limit after exponential back-off has exhausted, and they'll be returned as COORDINATOR_NOT_AVAILABLE to the client with an appropriate error message. This one needs adding to the KIP.

GROUP_ID_NOT_FOUND and UNKNOWN_TOPIC_OR_PARTITION mean that the share coordinator could not find the state records for these resources. This should never happen because the group coordinator will call the share coordinator to initialize the records for all share-partitions which can be assigned in share groups, thus meaning that the share coordinator will be able to find the state records for all partitions that can be assigned. These errors indicate a mismatch between the share-group state topic and what the client is requesting, and will be treated as a badly behaving client fetch data for share-partitions that it was never assigned. The error code will be INVALID_REQUEST.

FENCED_LEADER_EPOCH will cause the share-partition leader to recheck the leader epoch for the partition. If the leader epoch has changed underneath the share-partition leader, it needs to discard the share partition and retry, so this one is NOT_LEADER_OR_FOLLOWER. If it finds that the leader epoch has not changed but it's still being fenced, this is UNKNOWN_SERVER_ERROR.

Reply:
I have updated the PR with handling, also made the write persister RPC async. I have passed the errors from Persister to Share Partition Manager as there we can take the action of discarding the share partition or re-fetching the partition epoch. I ll handle that in subsequent PR once leader epoch handling is merged.